platform/darwin/src/MGLObserver_Private.h

```diff
@@ -0,0 +1,40 @@
+#import <mbgl/util/observable.hpp>
+
+#import "MGLObserver.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+#pragma mark - Native C++ peer object
+
+namespace mbgl {
+namespace darwin {
+
+class Observer : public mbgl::Observer {
+public:
+    Observer(MGLObserver *observer_): observer(observer_) {}
+    virtual ~Observer() = default;
+    virtual void notify(const ObservableEvent& event);
+    virtual std::size_t id() const;
+
+protected:
+
+    /// Cocoa observer that this adapter bridges to.
+    __weak MGLObserver *observer = nullptr;
```

Comment:
if `observer` is GCd, who will unsubscribe `mbgl::Observer` from the map?

Reply:
This is handled by documentation of `MGLObservable` - the expectation is that subscribe will retain the `MGLObserver` (and unsubscribe will release). This is done by storing the observer in a set.

The assert in the `MGLObserver` dealloc is for development, to help identify cases where we have neglected to retain the observers.

Another option that we might consider is to extend this, so that `MGLObserver` isn't actually public - and instead we use a notification handler. This would allow us to control `MGLObserver` completely, and would feel a little more idiomatic. (In fact, as you've noticed the base class already has a handler - this may or may not make it through to release).